src/measure_theory/pi_system.lean

```diff
@@ -268,9 +268,134 @@ begin
     apply h_b_h },
 end
 
+section Union_Inter
+
+/-! ### π-system generated by finite intersections of sets of a π-system family -/
+
+/-- From a set of finsets `S : set (finset ι)` and a family of sets of sets `π : ι → set (set α)`,
+define the set of sets that can be written as `⋂ x ∈ t, f x` for some `t ∈ S` and sets `f x ∈ π x`.
+
+If `S` is union-closed and `π` is a family of π-systems, then it is a π-system.
+The π-systems used to prove Kolmogorov's 0-1 law are of that form. -/
+def pi_Union_Inter {α ι} (π : ι → set (set α)) (S : set (finset ι)) : set (set α) :=
+{s : set α | ∃ (t : finset ι) (htS : t ∈ S) (f : ι → set α) (hf : ∀ x, x ∈ t → f x ∈ π x),
+  s = ⋂ x ∈ t, f x}
+
+/-- If `S` is union-closed and `π` is a family of π-systems, then `pi_Union_Inter π S` is a
+π-system. -/
+lemma is_pi_system_pi_Union_Inter {α ι} (π : ι → set (set α))
+  (hpi : ∀ x, is_pi_system (π x)) (S : set (finset ι)) (h_sup : sup_closed S) :
+  is_pi_system (pi_Union_Inter π S) :=
+begin
+  rintros t1 ⟨p1, hp1S, f1, hf1m, ht1_eq⟩ t2 ⟨p2, hp2S, f2, hf2m, ht2_eq⟩ h_nonempty,
+  simp_rw [pi_Union_Inter, set.mem_set_of_eq],
+  let g := λ n, (ite (n ∈ p1) (f1 n) set.univ) ∩ (ite (n ∈ p2) (f2 n) set.univ),
+  use [p1 ∪ p2, h_sup p1 p2 hp1S hp2S, g],
+  have h_inter_eq : t1 ∩ t2 = ⋂ i ∈ p1 ∪ p2, g i,
+  { rw [ht1_eq, ht2_eq],
+    simp_rw [← set.inf_eq_inter, g],
+    ext1 x,
+    simp only [inf_eq_inter, mem_inter_eq, mem_Inter, finset.mem_union],
+    refine ⟨λ h i hi_mem_union, _, λ h, ⟨λ i hi1, _, λ i hi2, _⟩⟩,
+    { split_ifs,
+      exacts [⟨h.1 i h_1, h.2 i h_2⟩, ⟨h.1 i h_1, set.mem_univ _⟩,
+        ⟨set.mem_univ _, h.2 i h_2⟩, ⟨set.mem_univ _, set.mem_univ _⟩], },
+    { specialize h i (or.inl hi1),
+      rw if_pos hi1 at h,
+      exact h.1, },
+    { specialize h i (or.inr hi2),
+      rw if_pos hi2 at h,
+      exact h.2, }, },
+  refine ⟨λ n hn, _, h_inter_eq⟩,
+  simp_rw g,
+  split_ifs with hn1 hn2,
+  { refine hpi n (f1 n) (hf1m n hn1) (f2 n) (hf2m n hn2) (set.ne_empty_iff_nonempty.mp (λ h, _)),
+    rw h_inter_eq at h_nonempty,
+    suffices h_empty : (⋂ i ∈ p1 ∪ p2, g i) = ∅,
+      from (set.not_nonempty_iff_eq_empty.mpr h_empty) h_nonempty,
+    refine le_antisymm (set.Inter_subset_of_subset n _) (set.empty_subset _),
+    refine set.Inter_subset_of_subset hn _,
+    simp_rw [g, if_pos hn1, if_pos hn2],
+    exact h.subset, },
+  { simp [hf1m n hn1], },
+  { simp [hf2m n h], },
+  { exact absurd hn (by simp [hn1, h]), },
+end
+
+lemma generate_from_pi_Union_Inter_le {α ι} {m : measurable_space α}
+  {s : ι → measurable_space α} (h : ∀ n, s n ≤ m) (π : ι → set (set α)) (S : set (finset ι))
```

Comment:
I am not sure you need `s`  here: you could just formulate your assumption `h` as : `∀ n, generate_from (π n) ≤ m`.